### PR TITLE
new task: scheduled price changes

### PR DESCRIPTION
**Pull request by tekhaus**
Given the complexity of this task, I would like a review of it's current (almost finished) state to make sure the choices made are sensible with regards to supporting a task like this in the public task library.

This task was developed based on a community task request, with several members chiming in on the features they would like, and what gaps they had with the LaaunchPad solution (Plus only) for scheduling price change events. Some additional features were gleamed from merchant posts in the community forums.

**Features:**
- Ability to schedule overlapping price change events
- Ability to cancel an ongoing event
- Ability to add collections that contain more than 1000 products
- Ability to discount more than 20k products
- Support percentage, fixed price, and fixed discount options for products *AND collections*
- Option to include specific SKUs in a list with a specific discount for the whole list
- Option to exclude SKUs from any discount qualification
- Option to exclude products by tag from any discount qualification
- Option to have (or not have) compare at prices set to original prices for the duration of the event

The primary complexity of this task is due to the decision to have it be triggered with **manual run keywords**, which allow a single task instance to handle scheduling, listing, and cancelling events. If the more traditional approach was used, then the task would essentially be scheduling new events on each save, and there would be no facility for cancelling an ongoing event. Cancelling a scheduled event that was some extended time in the future could also be problematic since the task event logs truncate after ~20 entries. The approach this task takes is to save each scheduled event and its status in a shop metafield.

Some of the additional complexity in this task is an attempt to workaround the limited task configuration option field types. There is extra validation in the task on the start and end date formats (which support dates and datetimes), on how discount percentages and positive and negative discount values are handled (fixed price and fixed discount respectively), and on manual keyword entries (which drive the task logic).

**Task development TODOS:**
- Write up the task docs, which will be lengthy due to explaining how this task operates in contrast to all the others, and for detailing the supported field formats (dates/datetimes, percentages, positive/negative floats)
- Add email notifications for scheduled, cancelled, and completed events, which includes the price change event configuration summary
- Decision on whether to auto-purge completed events from the shop metafield that stores all of the price change events, or whether this should be handled as a manual task keyword (e.g. "clean"). The cleaning would likely be needed for shops that run lots of price change events with many of the options configured, as there is a 65k size limit on JSON metafields.

*To be included in Docs*
Priority of discount assignment:
- Exclusions will always override inclusion:
  - Variant is already part of an ongoing price change event
  - SKU exclusion list
  - Product tag exclusion list
- SKU inclusion list
- Collection membership in order of configuration

### CODE REVIEW

File: docs/scheduled-price-changes/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+{
+  "event_start_datetime__required": null,
+  "event_end_datetime__required": null,
+  "set_compare_at_prices_to_original_price_during_event__boolean": null,
+  "collection_handles_and_discounts__keyval": null,
+  "skus_to_include__array": null,
+  "sku_discount": null,
+  "exclude_products_tagged_with__array": null,
+  "skus_to_exclude__array": null,
+  "send_task_notiication_emails_to__array_required": null,
```
Comment (isaacbowen) [inline docs/scheduled-price-changes/README.md:23]: typo

**Comment by isaacbowen** [comment]
I like this approach. :) Even the run keyword approach works for me. A future still exists for a native event type, in the style of mechanic/user/trigger, which allows for _predefined_ choices at runtime, and seeing a real use-case for that here is helpful.

A request: please add “Advanced: ” to the beginning of the task name here. We’re adding a new task .. category? type? delineation? allowing for tasks which we won’t support in the same way as the rest of the library. (This may end up being a replacement of the “Demonstration” label, which we’ve got on a bunch of tasks already.) Once you’ve got this merged in, I’ll follow up with some work on the library itself to clearly set expectations about what we’ll support (i.e. most of the library), and what will require external developer attention for support (i.e. tasks like this one).

File: docs/README.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 This directory is built automatically. Each task's documentation is generated from the corresponding task export file in [../tasks](../tasks/). To contribute, see [../CONTRIBUTING.md](../CONTRIBUTING.md).
 
+* [ADVANCED: Scheduled Price Changes](./advanced-scheduled-price-changes)
```
Comment (mattsodomsky) [inline docs/README.md:5]: Changes to regular title case (Advanced:)

File: docs/advanced-scheduled-price-changes/README.md [hunk 3]
```diff
@@ -1,8 +1,8 @@
-# Scheduled Price Changes
+# ADVANCED: Scheduled Price Changes
 
-Tags: Price, Products, Schedule
+Tags: Advanced, Price, Products, Schedule
 
-Run task manually, and enter a keyword of __schedule__, __list__, __email__, or __cancel__. The cancel keyword must be followed by the *price change event ID*, which will be generated when an event is successfully scheduled.
+This advanced task allows you to schedule price change events for your store, useful for seasonal and flash sales. To use this task, configure a price change event using the task options, save the task and run it manually, and then enter a keyword of **schedule** in the dialog that appears. An email summary with the price change event ID will be sent upon successful scheduling.
```
Comment (mattsodomsky) [inline docs/advanced-scheduled-price-changes/README.md:5]: Can you add this sentence:

🚨  This is an advanced task, intended for technical users. For tasks like these, the Mechanic team only offers support with platform-level issues. For help with task-level issues (e.g. debugging, configuration, customization, monitoring, etc), see [Hire a Mechanic developer](https://learn.mechanic.dev/hire-a-developer).

File: docs/advanced-scheduled-price-changes/README.md [hunk 4]
```diff
@@ -33,15 +32,56 @@ Run task manually, and enter a keyword of __schedule__, __list__, __email__, or
...
 ```
 
 [Learn about event subscriptions in Mechanic](https://docs.usemechanic.com/article/408-subscriptions)
 
 ## Documentation
 
-Run task manually, and enter a keyword of __schedule__, __list__, __email__, or __cancel__. The cancel keyword must be followed by the *price change event ID*, which will be generated when an event is successfully scheduled.
-
-## TODO: Docs
+This advanced task allows you to schedule price change events for your store, useful for seasonal and flash sales. To use this task, configure a price change event using the task options, save the task and run it manually, and then enter a keyword of **schedule** in the dialog that appears. An email summary with the price change event ID will be sent upon successful scheduling.
```
Comment (mattsodomsky) [inline docs/advanced-scheduled-price-changes/README.md:42]: Can you add this sentence:

🚨  This is an advanced task, intended for technical users. For tasks like these, the Mechanic team only offers support with platform-level issues. For help with task-level issues (e.g. debugging, configuration, customization, monitoring, etc), see [Hire a Mechanic developer](https://learn.mechanic.dev/hire-a-developer).

File: docs/advanced-scheduled-price-changes/README.md [hunk 5]
```diff
@@ -1,8 +1,8 @@
-# Scheduled Price Changes
+# ADVANCED: Scheduled Price Changes
 
-Tags: Price, Products, Schedule
+Tags: Advanced, Price, Products, Schedule
```
Comment (mattsodomsky) [inline docs/advanced-scheduled-price-changes/README.md:3]: Can you add tags: sale, discount

^^undestanding this isn't quite a discount, but I think it'll help with task discoverability

**Review by mattsodomsky — CHANGES_REQUESTED** [review]
Looks awesome, minor changes and we are good to publish

**Comment by isaacbowen** [comment]
@tekhaus I added a couple changes to this repo (#93, #94) that help with standardizing communication around "Advanced" tasks. Can you rebase (or merge master) before this pr gets wrapped up, so we can see how that looks with your changes?

**Review by mattsodomsky — APPROVED** [review]
Looks great, nice work!

**Pull request merged**